### PR TITLE
Fix Discord notifications not being sent due to templating issues

### PR DESCRIPTION
**Pull request by santihernandezc**
**What this PR does / why we need it**:

If template parsing fails for a Discord notification, the `Notify()` function errors out and the notification is never sent.

Screen Shot 2022-04-25 at 13 29 07

With this PR, templating errors don't prevent notifications from being sent, and errors are logged with a `Warn` level. This makes the Discord notifier consistent with what we do in other contact points like Teams, DingDing or VictorOps.

**Which issue(s) this PR fixes**:

There's no issue created for this specific fix, but it's related to https://github.com/grafana/support-escalations/issues/2261, where templating issues resulted in empty URLs and notifications not being sent.

### CODE REVIEW

File: pkg/services/ngalert/notifier/channels/discord_test.go [hunk 1]
```diff
@@ -58,6 +57,70 @@ func TestDiscordNotifier(t *testing.T) {
...
+					"title": "[FIRING:1]  (val1)",
+					"url":   "http://localhost/alerting/list",
+					"type":  "rich",
+				}},
+				"username": "Grafana",
+			},
+			expMsgError: nil,
+		},
+		{
+			name: "Invalid template",
```
Comment (alexweav) [inline pkg/services/ngalert/notifier/channels/discord_test.go:93]: The PR fixes 3 distinct error cases:
- when the template for `content` fails (should truncate and affect no other fields)
- when the template for `avatar_url` fails (should replace with default avatar and affect no other fields)
- when the template for `url` fails (should replace with default url)

But, the added tests only cover the first one - the tests only cover invalid templates in `content`. Could we please add additional test cases, using an invalid `avatar_url` and an invalid `url` template?

File: pkg/services/ngalert/notifier/channels/factory.go [hunk 2]
```diff
@@ -36,7 +36,7 @@ func NewFactoryConfig(config *NotificationChannelConfig, notificationService not
 var receiverFactories = map[string]func(FactoryConfig) (NotificationChannel, error){
 	"prometheus-alertmanager": AlertmanagerFactory,
 	"dingding":                DingDingFactory,
-	"discord":                 DiscrodFactory,
+	"discord":                 DiscordFactory,
```
Comment (alexweav) [inline pkg/services/ngalert/notifier/channels/factory.go:39]: Thanks for fixing this while you are in the area!

**Review by alexweav — COMMENTED** [review]
Looking good, couple nitpicky comments

File: pkg/services/ngalert/notifier/channels/discord.go [hunk 3]
```diff
@@ -119,10 +129,15 @@ func (d DiscordNotifier) Notify(ctx context.Context, as ...*types.Alert) (bool,
...
 	if tmplErr != nil {
 		d.log.Warn("failed to template Discord message", "err", tmplErr.Error())
-		return false, tmplErr
+		tmplErr = nil
+	}
+
+	u := tmpl(d.WebhookURL)
+	if tmplErr != nil {
+		d.log.Warn("failed to template Discord URL", "err", tmplErr.Error(), "fallback", d.WebhookURL)
+		u = d.WebhookURL
```
[the remark this answers is not in the document]
Comment (santihernandezc) [reply pkg/services/ngalert/notifier/channels/discord.go:140]: This error handling pattern is gonna be changed by @JacobsonMT in a future PR, so maybe it's not necessary to change this now.

**Review by alexweav — APPROVED** [review]
LGTM

**Pull request merged**